lib/common/balancer/hash.c

```diff
@@ -0,0 +1,312 @@
+    else return 0;
+}
+
+static size_t find_bucket_for_item(hash_bucket_vector_t *ring, void *key, size_t key_size)
+{
+    uint64_t hash = compute_hash(key, key_size);
+    size_t index = range_bsearch(&hash, ring->entries, ring->size, sizeof(struct hash_bucket_t),
+                                 hash_cmp, 1);
+    return index;
+}
+
+static void insert_new_bucket(hash_bucket_vector_t *ring, struct hash_bucket_t *bucket)
+{
+    size_t index = range_bsearch(bucket, ring->entries, ring->size, sizeof(struct hash_bucket_t),
+                                 hash_cmp, 0);
+    size_t i;
+
+    if (index != ring->size)
+        for (i = ring->size; i > index; i--)
+            ring->entries[i] = ring->entries[i - 1];
+    ring->entries[index] = *bucket;
+    ring->size++;
+}
+
+static void add_bucket(hash_bucket_vector_t *ring, size_t *request_count, const void *key, size_t key_len)
+{
+    struct hash_bucket_t bucket = {compute_hash(key, key_len), request_count};
+    if (ring->capacity == ring->size) {
+        h2o_vector_reserve(NULL, ring, 2 * ring->capacity);
+    }
+    insert_new_bucket(ring, &bucket);
+}
+
+static void construct(h2o_socketpool_target_vector_t *targets, void *_conf, void **data)
+{
+    struct bounded_hash_conf_t *conf = _conf;
+    size_t i;
+    struct bounded_hash_t *self = h2o_mem_alloc(sizeof(*self));
+    char tag_buf[NI_MAXHOST + sizeof(":65535")];
+    size_t tag_buf_len;
+
+    if (conf == NULL) {
+        self->c = 1.2;
+        self->type = H2O_BALANCER_HASH_KEY_TYPE_IP_PORT;
+    } else {
+        self->c = conf->c;
+        self->type = conf->type;
+    }
+    pthread_mutex_init(&self->mutex, NULL);
+    memset(&self->buckets, 0, sizeof(self->buckets));
+    
+    h2o_vector_reserve(NULL, &self->buckets, targets->size);
+    for (i = 0; i < targets->size; i++) {
+        tag_buf_len = sprintf(tag_buf, "%s:%zu", targets->entries[i]->url.host.base, i);
+        add_bucket(&self->buckets, &targets->entries[i]->_shared.leased_count, tag_buf, tag_buf_len);
+    }
+    *data = self;
+}
+
+static size_t bounded_find_bucket(hash_bucket_vector_t buckets, size_t startat, double c)
+{
+    size_t index = startat;
+    size_t i;
+    size_t total = 0;
+    double lower_total_bound = (1.0 / (c - 1)) * buckets.size;
+    for (i = 0; i < buckets.size; i++) {
+        total += *buckets.entries[i].request_count;
+    }
+    
+    double bound_total = c * (total + 1);
+    
+    if (total > lower_total_bound) {
+        for (i = 0; i < buckets.size; i++) {
+            if (bound_total >= buckets.size * (*buckets.entries[i].request_count + 1)) {
+                break;
+            }
+            if (++index == buckets.size) {
+                index = 0;
+            }
+        }
+    }
+    
+    assert(bound_total <= lower_total_bound || i != buckets.size);
+    
+    return index;
+}
+
+static size_t selector(h2o_socketpool_target_vector_t *targets, void *_data, int *tried, void *_req)
+{
+    h2o_req_t *req = _req;
```

Comment:
Could you please consider removing direct dependency to `h2o_req_t` from here? The general rule is that the files under lib/common (and the associated header files) cannot depend to files under lib/core or h2o.h.

Also, I am not sure if adding `req` as an argument (and also as a member visible in socketpool.h) is necessary. I would argue that it is the caller of the socketpool's responsibility to return any information required for load balancing. And the caller already has `void *data` passed as an argument to `h2o_socketpool_connect` as a stash.

Reply:
Thanks for your comment. It is possible to remove direct dependency to `h2o_req_t` but the argument is necessary when it comes to hash balancer. When using hash balancer, currently it support `ip`, `ip-port`, or `path` as hash key of a request, each needs some information of the request. So there should be some way to pass them.